File: 8-string-to-integer-atoi/8-string-to-integer-atoi.py
class Solution(object):
    def myAtoi(self, s):
        """
        :type s: str
        :rtype: int
        """
        dfa = [{'+-':1,'alpha':3,'whitespace':0,'digit':2,'.':3},
               {'digit':2,'+-':3,'alpha':3,'whitespace':3,'.':3},
               {'digit':2,'+-':3,'alpha':3,'whitespace':3,'.':3},
               {'digit':3,'+-':3,'alpha':3,'whitespace':3,'.':3}
            
        ]
        cur = 0
        sign = 1
        res = 0
        for char in s:
            if char in ['+','-']:
                group = '+-'
                if char == '-' and cur == 0:
                    sign = -1
            elif char == ' ':
                group = 'whitespace'
            elif 'a' <= char.lower() <= 'z':
                group = 'alpha'
            elif char == '.':
                group = '.'
            else:
                group = 'digit'
                res = 10*res + int(char)
            if group not in dfa[cur]:
                break
            cur = dfa[cur][group]
            if cur == 3:
                break
        if -2**31<= res*sign <= 2**31-1:
            return res*sign
        elif res*sign < -2**31:
            return -2**31
        else:
            return 2**31-1
            
                
        

File: 8-string-to-integer-atoi/8-string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {
        //remove leading whitespace
        s = s.replaceAll("^\\s+","");
        // the first digit
        if (s.length() == 0)
            return 0;
        
        int sign = 1;
        if (s.charAt(0) == '-')
            sign = -1;
        int res = 0;
        
        
        if (Character.isDigit(s.charAt(0))){
            res = Character.getNumericValue(s.charAt(0));
        }
        
        if(!Character.isDigit(s.charAt(0)) && s.charAt(0) != '-' && s.charAt(0) != '+')
            return 0;
        
        int pointer = 0;
        //read digit till non-digit

        s = s.substring(1);
        res = sign * res;
        while(pointer < s.length() && Character.isDigit(s.charAt(pointer))){
            int cur_num = Character.getNumericValue(s.charAt(pointer++));
            //negative overflow
            if (sign == -1){
            // not negative overflow
            if (res < Integer.MIN_VALUE/10 || (res == Integer.MIN_VALUE / 10 && cur_num > 8))
                return Integer.MIN_VALUE;
            res = 10 * res - cur_num;
            }
            else{
            
                if (res > Integer.MAX_VALUE/10 || (res == Integer.MAX_VALUE / 10 && cur_num > 7))
                return Integer.MAX_VALUE;
            res = res * 10 + cur_num;
            
            }
        }
        
        return res;
        
        
    }
}